File: lab789.cpp
#include <iostream>
#include <cmath>
#include <complex>
#include <vector>

using namespace std;

const double PI = 3.14159265;

typedef complex<double> Complex;
typedef vector<Complex> Vector;

Vector dft(Vector x) {
    size_t N = x.size();
    Vector A(N);

    for (size_t k = 0; k < N; k++) {
        A[k] = 0;
        for (size_t j = 0; j < N; j++) {
            double angle = 2 * PI * k * j / N;
            Complex e(cos(angle), -sin(angle));
            A[k] += x[j] * e;
        }
        A[k] /= N;
    }

    return A;
}

Vector idft(Vector x) {
    size_t N = x.size();
    Vector A(N);

    for (size_t k = 0; k < N; k++) {
        A[k] = 0;
        for (size_t j = 0; j < N; j++) {
            double angle = 2 * PI * k * j / N;
            Complex e(cos(angle), sin(angle));
            A[k] += x[j] * e;
        }
    }

    return A;
}

Vector sft(Vector x, size_t p1, size_t p2) {
    size_t N = x.size();
    Vector A1(N);

    for (size_t k1 = 0; k1 < p1; k1++)
        for (size_t j2 = 0; j2 < p2; j2++) {
            A1[k1 + j2 * p1] = 0;
            for (size_t j1 = 0; j1 < p1; j1++) {
                double angle = 2 * PI * j1 * k1 / p1;
                Complex e(cos(angle), -sin(angle));
                A1[k1 + j2 * p1] += x[j2 + p2 * j1] * e;
            }
            A1[k1 + j2 * p1] /= p1;
        }

    Vector A2(N);

    for (size_t k1 = 0; k1 < p1; k1++)
        for (size_t k2 = 0; k2 < p2; k2++) {
            A2[k1 + k2 * p1] = 0;
            for (size_t j2 = 0; j2 < p2; j2++) {
                double angle = 2 * PI * j2 / p1 / p2 * (k1 + p1 * k2);
                Complex e(cos(angle), -sin(angle));
                A2[k1 + k2 * p1] += A1[k1 + j2 * p1] * e;
            }
            A2[k1 + k2 * p1] /= p2;
        }

    return A2;
}

Vector rsft(Vector x, size_t p1, size_t p2) {
    size_t N = x.size();
    Vector A1(N);

    for (size_t k1 = 0; k1 < p1; k1++)
        for (size_t j2 = 0; j2 < p2; j2++) {
            A1[k1 + j2 * p1] = 0;
            for (size_t j1 = 0; j1 < p1; j1++) {
                double angle = 2 * PI * j1 * k1 / p1;
                Complex e(cos(angle), sin(angle));
                A1[k1 + j2 * p1] += x[j2 + p2 * j1] * e;
            }
        }

    Vector A2(N);

    for (size_t k1 = 0; k1 < p1; k1++)
        for (size_t k2 = 0; k2 < p2; k2++) {
            A2[k1 + k2 * p1] = 0;
            for (size_t j2 = 0; j2 < p2; j2++) {
                double angle = 2 * PI * j2 / p1 / p2 * (k1 + p1 * k2);
                complex<double> e(cos(angle), sin(angle));
                A2[k1 + k2 * p1] += A1[k1 + j2 * p1] * e;
            }
        }

    return A2;
}


int main() {
	Vector a(100), b(100);
    for (int i = 0; i < 100; i++) {
        a[i] = i;
        b[i] = i;
    }
	size_t N = a.size() + b.size() - 1, p1, p2;
	
    cout << fixed;
    cout.precision(4);

    cout << endl << "Transformation:" << endl;
    /*a = dft(a);
    b = dft(b);*/
    p1 = sqrt(a.size());
    p2 = sqrt(a.size());
    a = sft(a, p1, p2);
    p1 = sqrt(b.size());
    p2 = sqrt(b.size());
    b = sft(b, p1, p2);
    for (size_t i = 0; i < a.size(); i++)
        cout << a[i] << " ";
    cout << endl;
    for (size_t i = 0; i < b.size(); i++)
        cout << b[i] << " ";
    cout << endl;

    Vector c(N);
    for (size_t i = 0; i < N; i++) {
        for (size_t j = 0; j < a.size(); j++)
            for (size_t t = 0; t < b.size(); t++)
                if (j + t == i)
                    c[i] += a[j] * b[t];
        cout << c[i] << endl;
    }

    cout << endl << "Reverse:" << endl;
  /*  a = idft(a);
    b = idft(b);*/
    p1 = sqrt(a.size());
    p2 = sqrt(a.size());
    a = rsft(a, p1, p2);
    p1 = sqrt(b.size());
    p2 = sqrt(b.size());
    b = rsft(b, p1, p2);
    for (size_t i = 0; i < a.size(); i++)
        cout << a[i] << " ";
    cout << endl;
    for (size_t i = 0; i < b.size(); i++)
        cout << b[i] << " ";
    cout << endl;

	for (size_t i = 0; i < N; i++) {
        c[i] = 0;
		for (size_t j = 0; j < a.size(); j++)
			for (size_t t = 0; t < b.size(); t++)
				if (j + t == i)
					c[i] += a[j] * b[t];
		cout << c[i] << endl;
	}

	return 0;
}

File: lab789.py
import math
import cmath

PI = 3.14159265

# Функция DFT (дискретное преобразование Фурье)
def dft(x):
    N = len(x)
    A = [0] * N
    for k in range(N):
        A[k] = 0
        for j in range(N):
            angle = 2 * PI * k * j / N
            e = cmath.exp(-1j * angle)
            A[k] += x[j] * e
        A[k] /= N
    return A

# Обратное DFT (IDFT)
def idft(x):
    N = len(x)
    A = [0] * N
    for k in range(N):
        A[k] = 0
        for j in range(N):
            angle = 2 * PI * k * j / N
            e = cmath.exp(1j * angle)
            A[k] += x[j] * e
    return A

# Функция SFT (преобразование в двух измерениях)
def sft(x, p1, p2):
    N = len(x)
    A1 = [0] * N

    for k1 in range(p1):
        for j2 in range(p2):
            idx1 = k1 + j2 * p1
            A1[idx1] = 0
            for j1 in range(p1):
                angle = 2 * PI * j1 * k1 / p1
                e = cmath.exp(-1j * angle)
                A1[idx1] += x[j2 + p2 * j1] * e
            A1[idx1] /= p1

    A2 = [0] * N
    for k1 in range(p1):
        for k2 in range(p2):
            idx1 = k1 + k2 * p1
            A2[idx1] = 0
            for j2 in range(p2):
                angle = 2 * PI * j2 / p1 / p2 * (k1 + p1 * k2)
                e = cmath.exp(-1j * angle)
                A2[idx1] += A1[k1 + j2 * p1]
            A2[idx1] /= p2
    return A2

# Обратное SFT (RSFT)
def rsft(x, p1, p2):
    N = len(x)
    A1 = [0] * N

    for k1 in range(p1):
        for j2 in range(p2):
            idx1 = k1 + j2 * p1
            A1[idx1] = 0
            for j1 in range(p1):
                angle = 2 * PI * j1 * k1 / p1
                e = cmath.exp(1j * angle)
                A1[idx1] += x[j2 + p2 * j1]

    A2 = [0] * N
    for k1 in range(p1):
        for k2 in range(p2):
            idx1 = k1 + k2 * p1
            A2[idx1] = 0
            for j2 in range(p2):
                angle = 2 * PI * j2 / p1 / p2 * (k1 + p1 * k2)
                e = cmath.exp(1j * angle)
                A2[idx1] += A1[k1 + j2 * p1]
    return A2

# Основная программа
def main():
    a = [complex(i, 0) for i in range(100)]
    b = [complex(i, 0) for i in range(100)]
    
    N = len(a) + len(b) - 1
    
    print("Transformation:")
    # Преобразование SFT
    p1 = int(math.sqrt(len(a)))
    p2 = p1
    a = sft(a, p1, p2)

    p1 = int(math.sqrt(len(b)))
    p2 = p1
    b = sft(b, p1, p2)

    print(a)
    print(b)

    # Вычисление свертки векторов
    c = [0] * N
    for i in range(N):
        for j in range(len(a)):
            for t in range(len(b)):
                if j + t == i:
                    c[i] += a[j] * b[t]
        print(c[i])

    print("\nReverse:")

    # Обратное преобразование RSFT
    p1 = int(math.sqrt(len(a)))
    p2 = p1
    a = rsft(a, p1, p2)

    p1 = int(math.sqrt(len(b)))
    p2 = p1
    b = rsft(b, p1, p2)

    print(a)
    print(b)

    # Повторное вычисление свертки
    for i in range(N):
        c[i] = 0
        for j in range(len(a)):
            for t in range(len(b)):
                if j + t == i:
                    c[i] += a[j] * b[t]
        print(c[i])

if __name__ == "__main__":
    main()
